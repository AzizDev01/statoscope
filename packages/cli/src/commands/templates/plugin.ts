import { concatTemplateParts, templatePartsByOptions } from './templateParts';
import { FileExt, RuleTemplatesPartsFromFileExt, Template } from './types';

export const pluginTemplate: Template = (templateOptions) => {
  const plugin = templatePartsByOptions(templateOptions, {
    js: jsPlugin,
    ts: tsPlugin,
  });

  return concatTemplateParts(plugin);
};

const jsPlugin: RuleTemplatesPartsFromFileExt[FileExt.js] = {
  imports: {
    commonjs: `const asyncChunkRule = require('./rule.statoscope.js')`,
    esm: `import asyncChunkRule from './rule.statoscope.js';`,
  },
  body: `/**
  * @typedef {import('@statoscope/webpack-model').Prepared} Prepared
  * @returns {import('@statoscope/stats-validator/dist/plugin').PluginDescription<Prepared>} PluginDescription
 */
const newStatoscopePlugin = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: {
    commonjs: 'module.exports = newStatoscopePlugin',
    esm: 'export default newStatoscopePlugin',
  },
};

const tsPlugin: RuleTemplatesPartsFromFileExt[FileExt.ts] = {
  imports: `import type { PluginFn } from '@statoscope/stats-validator/dist/plugin';
import type { Prepared } from '@statoscope/webpack-model';
import asyncChunkRule from './rule.statoscope.ts';`,
  body: `
const newStatoscopePlugin: PluginFn<Prepared> = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: 'export default newStatoscopePlugin',
};
