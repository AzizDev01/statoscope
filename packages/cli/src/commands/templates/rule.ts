export function ruleTemplate(): string {
  return `import { Rule } from '@statoscope/stats-validator/dist/rule';
import { Prepared } from '@statoscope/webpack-model';

//Statoscope Plugins examples:
https://github.com/statoscope/statoscope/tree/c9d0c148c1e38a87696d79485f37e598110db80c/packages/stats-validator-plugin-webpack#rules


${asyncChunkRule}

export default asyncChunkRule;`;
}

export const asyncChunkRule = `
type WebpackRule<TParams> = Rule<TParams, Prepared>;

type Params = string[];

const asyncChunkRule: WebpackRule<Params> = (ruleParams, data, api): void => {
  const normalizedParams: string[] = ruleParams ? ruleParams : [];
  const chunks = Array.from(new Set(...data.compilations.map((c) => c.chunks)));
  chunks
    .filter((c) =>
      c.names.some((n) => normalizedParams.includes(n) && c.initial !== false)
    )
    .forEach((f) => {
      api.message(f.names[0] + ' is not async chunk.');
    });
};`;
