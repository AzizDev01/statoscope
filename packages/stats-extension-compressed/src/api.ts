import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Resource, Size } from './generator';

export type API = (compilationId: string, resourceId: string) => Size | null;

const makeAPI: APIFactory<Format, API> = (source) => {
  const sizeResolvers: Map<string, Resolver<string, Resource>> = new Map();

  for (const compilation of source.payload.compilations) {
    sizeResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.resources, (r) => r.id)
    );
  }

  return (compilationId: string, resourceId: string): Size | null => {
    return sizeResolvers.get(compilationId)?.(resourceId)?.size ?? null;
  };
};

export default makeAPI;
