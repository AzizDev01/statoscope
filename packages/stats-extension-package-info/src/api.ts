import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Package, Instance } from './generator';

export type API = {
  getPackage: (compilationId: string, packageName: string) => Package | null;
  getInstance: (
    compilationId: string,
    packageName: string,
    instancePath: string
  ) => Instance | null;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const packageResolvers: Map<string, Resolver<string, Package>> = new Map();
  const instanceResolvers: Map<Package, Resolver<string, Instance>> = new Map();

  for (const compilation of source.payload.compilations) {
    packageResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.packages, (r) => r.name)
    );

    for (const packageItem of compilation.packages) {
      instanceResolvers.set(
        packageItem,
        makeEntityResolver(packageItem.instances, (r) => r.path)
      );
    }
  }

  return {
    getPackage(compilationId: string, packageId: string): Package | null {
      return packageResolvers.get(compilationId)?.(packageId) ?? null;
    },
    getInstance(
      compilationId: string,
      packageId: string,
      instancePath: string
    ): Instance | null {
      const resolvedPackage = packageResolvers.get(compilationId)?.(packageId) ?? null;

      if (!resolvedPackage) {
        return null;
      }

      return instanceResolvers.get(resolvedPackage)?.(instancePath) ?? null;
    },
  };
};

export default makeAPI;
