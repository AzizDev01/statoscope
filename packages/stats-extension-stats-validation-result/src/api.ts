import makeEntityResolver from '@statoscope/helpers/dist/entity-resolver';
import { Resolver } from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { RelatedItem, RuleDescriptor } from '@statoscope/types/types/validation';
import { Format, Item } from './generator';

export type API = {
  getItems: (
    compilationId: string | null,
    type?: RelatedItem['type'] | null,
    relatedId?: string | number
  ) => Item[];
  getItemById(id: number): Item | null;
  getRule(id: string): RuleDescriptor | null;
};

type IndexItem = { item: Item; related: RelatedItem };
type Index = {
  type: RelatedItem['type'];
  links: IndexItem[];
  items: Item[];
  related: RelatedItem[];
  byRelatedId: Resolver<string | number, Item>;
};

export type IndexAPI = {
  readonly type: RelatedItem['type'];
  add(item: Item, related: RelatedItem): void;
  getItems(id?: string | number): Item[];
};

function makeIndex(type: RelatedItem['type']): IndexAPI {
  const links: IndexItem[] = [];
  const index: Index = {
    type,
    links,
    items: [],
    related: [],
    byRelatedId: makeEntityResolver(
      links,
      (item) => item.related.id,
      (item) => item.item
    ),
  };

  return {
    type,
    add(item: Item, related: RelatedItem): void {
      if (!index.items.includes(item)) {
        index.items.push(item);
      }

      if (!index.related.includes(related)) {
        index.related.push(related);
      }

      if (!index.links.find((link) => link.item === item && link.related === related)) {
        index.links.push({ item, related });
      }
    },
    getItems(id?: string | number): Item[] {
      if (id != null) {
        return (
          index.links
            // eslint-disable-next-line eqeqeq
            .filter((link) => link.related.id == id)
            .map((link) => link.item)
        );
      }

      return index.items;
    },
  };
}

export type CompilationData = {
  id: string | null;
  items: Item[];
  resolveIndex: Resolver<RelatedItem['type'], IndexAPI>;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const compilationResolvers: CompilationData[] = [];
  const resolveCompilationsResolvers = makeEntityResolver(
    compilationResolvers,
    (item) => item.id
  );
  const resolveRule = makeEntityResolver(
    source.payload.rules,
    (item) => item.name,
    (item) => item.descriptor
  );
  const items: Item[] = [];
  const resolveItemById = makeEntityResolver(items, (item) => item.id);
  for (const compilation of source.payload.compilations) {
    const indexes: IndexAPI[] = [];
    const compilationData: CompilationData = {
      id: compilation.id,
      items: compilation.items,
      resolveIndex: makeEntityResolver(indexes, (index) => index.type),
    };
    compilationResolvers.push(compilationData);

    for (const item of compilation.items) {
      items.push(item);
      for (const related of item.related) {
        let index = compilationData.resolveIndex(related.type);

        if (!index) {
          index = makeIndex(related.type);
          indexes.push(index);
        }

        index.add(item, related);
      }
    }
  }

  return {
    getItems: (
      compilationId: string | null,
      type?: RelatedItem['type'] | null,
      relatedId?: string | number
    ): Item[] => {
      if (relatedId) {
        if (!type) {
          throw new Error('type must be specified');
        }

        return (
          resolveCompilationsResolvers(compilationId)
            ?.resolveIndex(type)
            ?.getItems(relatedId) ?? []
        );
      }

      if (type) {
        return (
          resolveCompilationsResolvers(compilationId)?.resolveIndex(type)?.getItems() ??
          []
        );
      }

      return resolveCompilationsResolvers(compilationId)?.items ?? [];
    },
    getRule(id: string): RuleDescriptor | null {
      return resolveRule(id);
    },
    getItemById(id: number): Item | null {
      return resolveItemById(id);
    },
  };
};

export default makeAPI;
