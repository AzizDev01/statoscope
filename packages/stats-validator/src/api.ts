import {
  API,
  MakeAPIParams,
  RuleDescriptor,
  Storage,
} from '@statoscope/types/types/validation';

export function makeAPI(params?: MakeAPIParams): API {
  const storage: Storage = [];
  let hasErrors = false;
  let errors = 0;
  let warnings = 0;
  let infos = 0;
  let descriptor: RuleDescriptor | null = null;
  return {
    setRuleDescriptor(desc: RuleDescriptor): void {
      descriptor = desc;
    },
    getRuleDescriptor(): RuleDescriptor | null {
      return descriptor;
    },
    hasErrors(): boolean {
      return hasErrors;
    },
    getInfoTotal(): number {
      return infos;
    },
    getWarnTotal(): number {
      return warnings;
    },
    getErrorTotal(): number {
      return errors;
    },
    getStorage(): Storage {
      return storage;
    },
    warn(message, filenameOrOptions): void {
      if (params?.warnAsError) {
        hasErrors = true;
      }

      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      storage.push({
        type: 'warn',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        related: filenameOrOptions?.related,
        message,
      });
      warnings++;
    },
    error(message, filenameOrOptions): void {
      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      hasErrors = true;
      storage.push({
        type: 'error',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        related: filenameOrOptions?.related,
        message,
      });
      errors++;
    },
    info(message, filenameOrOptions): void {
      filenameOrOptions =
        typeof filenameOrOptions === 'string'
          ? { filename: filenameOrOptions }
          : filenameOrOptions;

      storage.push({
        type: 'info',
        filename: filenameOrOptions?.filename,
        compilation: filenameOrOptions?.compilation,
        details: filenameOrOptions?.details,
        related: filenameOrOptions?.related,
        message,
      });
      infos++;
    },
  };
}
