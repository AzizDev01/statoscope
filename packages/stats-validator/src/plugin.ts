import { Rule } from './rule';

export type InputFile = {
  name: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  data: any;
};

export type PrepareFn<TPrepared> = (files: InputFile[]) => TPrepared;
export type PluginDescription<TPrepared> = {
  prepare?: PrepareFn<TPrepared>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  rules?: Record<string, Rule<any, any>>;
};

export type PluginFn<TPrepared> = () => PluginDescription<TPrepared>;
