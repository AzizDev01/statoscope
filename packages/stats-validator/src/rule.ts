import { API } from './api';

export type RuleDataInput<TInput> = { reference?: TInput; input: TInput };

export type Rule<TParams, TInput> = (
  params: TParams | null,
  data: RuleDataInput<TInput>,
  api: API
) => void;
