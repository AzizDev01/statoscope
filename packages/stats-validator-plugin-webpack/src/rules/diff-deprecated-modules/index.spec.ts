import { makeAPI } from '@statoscope/stats-validator/dist/api';
import { RuleDataInput } from '@statoscope/stats-validator/dist/rule';
import { Prepared } from '@statoscope/webpack-model';
import plugin from '../..';
import statsV4 from '../../../../../test/bundles/v4/simple/stats-prod.json';
import statsV5 from '../../../../../test/bundles/simple/stats-prod.json';
import rule from './';

test('matches', () => {
  const pluginInstance = plugin();
  const prepared: RuleDataInput<Prepared> = pluginInstance.prepare!([
    { name: 'input.json', data: statsV5 },
    { name: 'reference.json', data: statsV4 },
  ]);
  const api = makeAPI({ warnAsError: false });

  rule([{ name: /.+/ }], prepared, api);

  expect(api.getStorage()).toMatchSnapshot();
});

test('not matches', () => {
  const pluginInstance = plugin();
  const prepared: RuleDataInput<Prepared> = pluginInstance.prepare!([
    { name: 'input.json', data: statsV5 },
    { name: 'reference.json', data: statsV4 },
  ]);
  const api = makeAPI({ warnAsError: false });

  rule([{ name: /not_exists/ }], prepared, api);

  expect(api.getStorage()).toMatchSnapshot();
});
