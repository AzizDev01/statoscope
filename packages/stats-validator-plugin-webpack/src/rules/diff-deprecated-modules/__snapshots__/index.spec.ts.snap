// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches 1`] = `
Array [
  Object {
    "compilation": undefined,
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "after": 1,
            "before": 0,
            "module": "./src/index.ts + 3 modules",
          },
        },
        "query": "
              $input: resolveInputFile();
              {
                module: $input.compilations.hash.(#.module.resolveModule($)).pick(),
                before: #.before,
                after: #.after,
              }
              ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Usage of ./src/index.ts + 3 modules was increased from 0 to 1",
    "related": Array [
      Object {
        "id": "./src/index.ts + 3 modules",
        "type": "module",
      },
    ],
  },
  Object {
    "compilation": undefined,
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "after": 4,
            "before": 0,
            "module": "../node_modules/is-array/index.js",
          },
        },
        "query": "
              $input: resolveInputFile();
              {
                module: $input.compilations.hash.(#.module.resolveModule($)).pick(),
                before: #.before,
                after: #.after,
              }
              ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Usage of ../node_modules/is-array/index.js was increased from 0 to 4",
    "related": Array [
      Object {
        "id": "../node_modules/is-array/index.js",
        "type": "module",
      },
    ],
  },
  Object {
    "compilation": undefined,
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "after": 2,
            "before": 0,
            "module": "../node_modules/css-loader/dist/runtime/api.js",
          },
        },
        "query": "
              $input: resolveInputFile();
              {
                module: $input.compilations.hash.(#.module.resolveModule($)).pick(),
                before: #.before,
                after: #.after,
              }
              ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Usage of ../node_modules/css-loader/dist/runtime/api.js was increased from 0 to 2",
    "related": Array [
      Object {
        "id": "../node_modules/css-loader/dist/runtime/api.js",
        "type": "module",
      },
    ],
  },
  Object {
    "compilation": undefined,
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "after": 2,
            "before": 0,
            "module": "../node_modules/css-loader/dist/runtime/getUrl.js",
          },
        },
        "query": "
              $input: resolveInputFile();
              {
                module: $input.compilations.hash.(#.module.resolveModule($)).pick(),
                before: #.before,
                after: #.after,
              }
              ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Usage of ../node_modules/css-loader/dist/runtime/getUrl.js was increased from 0 to 2",
    "related": Array [
      Object {
        "id": "../node_modules/css-loader/dist/runtime/getUrl.js",
        "type": "module",
      },
    ],
  },
  Object {
    "compilation": undefined,
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "after": 2,
            "before": 1,
            "module": "./node_modules/foo/node_modules/bar/index.js",
          },
        },
        "query": "
              $input: resolveInputFile();
              {
                module: $input.compilations.hash.(#.module.resolveModule($)).pick(),
                before: #.before,
                after: #.after,
              }
              ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Usage of ./node_modules/foo/node_modules/bar/index.js was increased from 1 to 2",
    "related": Array [
      Object {
        "id": "./node_modules/foo/node_modules/bar/index.js",
        "type": "module",
      },
    ],
  },
]
`;

exports[`not matches 1`] = `Array []`;
